Feat: Add Postgres database support + Refactor database handling

SakuraIsayeki opened the pull request:
<!-- Warning: If you create a pull request and wish to remain anonymous, you are highly advised to use Tails (https://tails.boum.org/) or a fresh git environment. We will *not* be able to help with anonymization after your pull request has been created. -->

<!-- Put the text you want in the changelog in your PR body so we can add it to the changelog. -->
**Added initial support for PostgreSQL database, as an alternative to MySQL or SQLite.** 
Foundational database elements have been refactored, to account for new implementation. Should actually be easier to work with, if another DB is to be be supported in the future.

Most of the initial flow has been tested, with register/login, groups, warps, being known good, regions still being tested. Bugs may arise with the new DB support, testing would be advised. 

Leaving this PR as a draft until I clear most of the bugs (which are mostly if not wholly due to casing and escaping issues), would be happy to hear some feedback in the meantime.

SakuraIsayeki committed:
refactor(server/db): Move database connection logic to separate class

Implements a DbBuilder class to streamline the creation of database connections for both SQLite and MySQL storage types.

Enhances error handling for database setup and ensures that necessary directories are created dynamically based on configuration settings.

This refactor improves code maintainability and readability, consolidating connection logic into a dedicated builder class.

SakuraIsayeki committed:
refactor(db): Move DbQueryBuilders to separate namespace

Removes obsolete query builder interface and implementations.

Introduces a new namespace for query-related classes to enhance organization.

Updates various database managers to use the new query utilities, promoting code reuse and maintainability.

Enhances query functionality, ensuring better consistency across the codebase.

SakuraIsayeki committed:
feat(db): Add PostgreSQL query creator implementation

Implements a query creator for PostgreSQL, allowing for proper database type mapping and table creation functionalities.

Enhances the library's database support by incorporating PostgreSQL-specific features such as SERIAL/BIGSERIAL for auto-increment columns and improved escape handling for table names.

Improves the extensibility and compatibility of the database framework with different SQL databases.

SakuraIsayeki committed:
feat(db): Add Postgres to SQL connection types + Refactor matching logic

Implements pattern matching for easier identification of database types.

Adds support for Postgres alongside existing Sqlite and MySQL types, enhancing flexibility for database connections.

Updates enum to include Postgres type.

SakuraIsayeki committed:
refactor(db): Update SqlQueryBuilder references + Various refactors

Consolidates the creation of SQL query builders across multiple classes to ensure a unified approach for database operations.

Replaces manual type checks and specific query creators with a generic method for better maintainability and to prevent errors.

Improves code readability and reduces duplication, facilitating easier updates in the future.

SakuraIsayeki committed:
feat(db): add Postgres support to configuration

Extends database configuration to support Postgres in addition to existing SQLite and MySQL options.

Includes new settings for Postgres host, database name, username, and password.

Implements a connection builder for Postgres, ensuring proper error handling when connecting.

Updates dependency to include Npgsql for Postgres connectivity.

SakuraIsayeki committed:
fix(db/postgres): Resolve SQL identifier casing issues

Improves SQL query execution by ensuring proper casing for identifiers across various database types, particularly for Postgres.

Enhances security and compatibility by using an identifier escaping method, preventing potential errors due to case sensitivity in SQL queries.

Addresses potential issues with existing queries for better reliability and consistency.

SakuraIsayeki committed:
fix(db): Correct casing and escaping in DB queries

Updates the database queries to handle casing inconsistencies
and improves SQL query parameter escaping for better
security and compatibility.

Refactors group existence checks for simplicity, enhancing
readability and maintainability.

Addresses issues related to unique constraints in user
registration by improving error handling for duplicate
usernames.

SakuraIsayeki committed:
fix(db): Fix casing for regions and warps

hakusaro reviewed (commented):
A very cursory glance from me indicates no major issues with this, but I wanna rope in @QuiCM who has more thoughts on database structure before I go too far.

SakuraIsayeki commented:
Thanks for the feedback. Indeed, more thoughts on the DB specialist(s) of the project would be greatly appreciated.
I was unsure of whether further evolutions on the database handling would be appropriate, and how they could/would take form. For now, decided to do the bare minimum to add support for PGSQL, all whilst keeping the structure as close to original.
I'm willing to go a few steps further, if you're okay with that. But further discussions may be needed. Hit me up on Discord if you have some time, and let's discuss it.

SakuraIsayeki committed:
refactor(db/pgsql): Revert SQL queries, lower column casing for PGSQL

Reverts SQL query identifier escaping to simplify queries and improves overall readability by using lowercase column names.

Enhances maintainability and alignment with database conventions by adopting a uniform casing scheme across all SQL operations.

Removes unnecessary complexity in query construction, streamlining the database operations performed within the application.

SakuraIsayeki committed:
Merge branch 'general-devel' into feature/database

hakusaro commented:
@Pryaxis/tshock I intend to merge this without any other feedback in **one week**.

SignatureBeef commented:
doing a bit of a quick test, unfortunately getting some errors but not ruling out something my end.
posting the exception should it be known:
![image](https://github.com/user-attachments/assets/87bc5294-d505-401a-b683-ea9193c12d6b)

running via launcher with `PostgreSQL 15.8 on aarch64-unknown-linux-gnu, compiled by gcc (GCC) 13.2.0, 64-bit`

aside from these that may need ironing out, the code changes are looking good to me, and personally happy to see this one pop up - nice stuff!.

SignatureBeef commented on TShockAPI/DB/SqlTable.cs:
```diff
@@ -69,36 +72,50 @@ public bool EnsureTableStructure(SqlTable table)
+				case SqlType.Postgres:
+				{
+					// HACK: Using "ilike" op to ignore case, due to weird case issues adapting for pgsql
+					using QueryResult reader = database.QueryReader("SELECT column_name FROM information_schema.columns WHERE table_name ILIKE @0", table.Name);
```
ah think we might need the table schema check here too, e.g. `AND table_schema=current_schema()` or similar which seems to work for me.
im currently utilising supabase's postgres which already has `auth.users`

SakuraIsayeki replied:
Fixed in latest commit, thanks for pointing that oversight out!

SignatureBeef reviewed (commented):
think i have found it - partially me, but think this might be a good case to check the table schema (e.g. in code review comment)

after this, seems that other error is resolved: 🎉 

![image](https://github.com/user-attachments/assets/4d642db0-a34f-49a0-8b00-769df2d31578)

SakuraIsayeki committed:
feat: Add schema filtering for Postgres table col scanning

Modifies the query to ensure it correctly retrieves column names by including the schema check.

hakusaro commented on TShockAPI/Configuration/TShockConfig.cs:
```diff
@@ -552,6 +552,22 @@ public class TShockSettings
 		public string MySqlPassword = "";
 
+ 		///<summary>The Postgres hostname and port to direct connections to.</summary>
+ 		[Description("The Postgres hostname and port to direct connections to.")]
```
Can you fix the inconsistent indendation / whitespace here?

SakuraIsayeki committed:
Merge branch 'general-devel' into feature/database

hakusaro commented:
@SakuraIsayeki I think you can rebase nao

SakuraIsayeki committed:
Merge branch 'general-devel' into feature/database

SakuraIsayeki commented:
> @SakuraIsayeki I think you can rebase nao

Rebased as requested. However, just before we merge anything, I'll quickly bump the Npgsql package too, to align with our current .NET version.

SakuraIsayeki committed:
build: Update Npgsql package version to 9.0.3 in project files

hakusaro commented on TShockAPI/Configuration/TShockConfig.cs:
```diff
@@ -552,6 +552,22 @@ public class TShockSettings
+
+	    /// <summary>The username used when connecting to a Postgres database.</summary>
+	    [Description("The username used when connecting to a Postgres database.")]
+	    public string PostgresUsername = "";
```
The inconsistent spacing/indentation in this file will be a problem later. While my previous commit is only on one line, all changes in this file have the problem.

hakusaro commented on TShockAPI/DB/DbBuilder.cs:
```diff
@@ -0,0 +1,115 @@
+/// </summary>
+public sealed class DbBuilder
+{
+	private readonly TShock _caller;
```
The indentation and spacing in this file doesn't seem to match the rest of the repo but I could be wrong. Can you double check?

SakuraIsayeki committed:
style: Correct indentation and formatting

Adjusts comment indentation and aligns property attribute formatting for clarity and consistency in configuration definitions.

SakuraIsayeki committed:
chore: Remove invalid comment on Npgsql package reference

SakuraIsayeki committed:
refactor: Rename query builder classes for consistency

Standardizes class names for different database query builders, aligning naming conventions across SQLite, MySQL, PostgreSQL implementations, and updating related factory method calls to improve code clarity and maintainability.

hakusaro commented:
@SakuraIsayeki please update the changelog on the wiki, kthx!
